File: 15. Tree/06ImplementationofPreorderTraversal.cpp
#include <bits/stdc++.h> 
using namespace std;

struct node {
    int key;
    node *left;
    node *right;

    node(int k) {
        key = k;
        left = right = NULL;
    }
};

void inorder(node *root){
    if(root!=NULL){
        inorder(root->left);
        cout<<root->key<<" ";
        inorder(root->right);
    }
}

int main() {
    node *root = new node(3);

    root->left = new node(2);
    root->right = new node(4);
    root->right->right = new node(5);
    root->left->left = new node(1);

    cout << "InOrder Traversal : "<<endl;

    inorder(root);

    return 0; 
}

File: 15. Tree/Python Visualizer for Trees/06ImplementationofPreorderTraversal.py
from visualiser.visualiser import Visualiser as vs

class Node:
    def __init__(self, key):
        self.key = key
        self.left = None
        self.right = None

@vs(node_properties_kwargs={"shape": "record", "color": "#f57542", "style": "filled", "fillcolor": "grey"})
def preorder(root):
    if root is not None:
        preorder(root.left)
        print(root.key, end=' ')
        preorder(root.right)

def create_sample_tree():
    root = Node(3)
    root.left = Node(2)
    root.right = Node(4)
    root.right.right = Node(5)
    root.left.left = Node(1)
    return root

def main():
    root = create_sample_tree()
    print("InOrder Traversal:")
    preorder(root)
    print()  # For a newline after traversal output
    vs.make_animation("preorder_traversal.gif", delay=5)

if __name__ == "__main__":
    main()
